fulfill

SheepTester opened the pull request:
<!-- Fill in all spots surrounded by brackets and confirm all check boxes after confirming completion of the tasks -->

# Info

Closes #225 

<!-- If there is no issue for this pull request yet, please create one or
delete this line if the pull request is for a very minor tweak. -->

fulfills request for fulfill button

<!-- What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context. -->

## Changes

- fulfill button for each order
- checkboxes for items that can be deselected to mark as not fulfilled
- got rid of fulfill view

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [ ] on the live deployment preview on Desktop.
- [ ] on the live deployment preview on Mobile.
- [ ] I have added new Cypress tests that are passing.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented any new functions in `/src/lib/*` and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

<!-- If you made any visual changes to the website, please include relevant screenshots below. -->

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/81173e64-4cea-4d86-9107-da0d023b2382)


![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/015bd2eb-aa22-4a7d-96fd-cdf08a6090dc)

SheepTester committed:
Add fulfill button

SheepTester committed:
Remove Fulfill view and move fulfill buttons to prepare view

SheepTester committed:
Do not show fulfill button if it's already fulfilled

SheepTester committed:
make order status component

SheepTester committed:
hide fulfill button if pickup event hasnt started yet

SheepTester committed:
Apparently partially fulfilled means they are missing some items

SheepTester committed:
move order row to separate component

SheepTester committed:
fulfill selected

SheepTester committed:
Merge branch 'main' into sean/fulfill

SheepTester committed:
fix type errors

alexzhang1618 commented on src/components/admin/store/PickupOrder/index.tsx:
```diff
@@ -0,0 +1,90 @@
+                  items: order.items.map(item =>
+                    itemUuids.includes(item.uuid) ? { ...item, fulfilled: true } : item
+                  ),
+                });
```
seeking to learn: does the API response not just return the whole new order updated correctly? do we need to manually set fulfilled for the items we fulfilled

SheepTester replied:
the API response doesn't return the `items` array so it won't say which items were fulfilled

alexzhang1618 reviewed (commented):
Code looks good! Good call on removing the order fulfill view i was just parroting off of v1 even though we don't really need it

Some usability comments:

can we put some gap between the status and fulfill button and maybe make the button the same size as the fulfill tag
<img width="1438" alt="Screenshot 2024-05-10 at 3 37 34 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/1c9b817c-074b-4f6e-91f0-d8d3f9779ec5">

also i think it would be nice to sort by status (i.e. put orders that still need action at the top)

I think we discussed this but I think ux wise it makes sense for all checkboxes to start blank and the merch person will tick each box off as they get the item

Another thing to consider: is it better to put the fulfill button under the merch items for each user? I think it's nicer so your mouse only has to stay on one side but idk

SheepTester committed:
Merge remote-tracking branch 'origin/main' into sean/fulfill

SheepTester committed:
always show fulfilled items

SheepTester committed:
sorting things and other misc changes:

- show green "fulfilled" when fulfilling a previously partially fulfilled order
- sort orders by PLACED first
- sort items alphabetically
- item breakdown: ignore fulfilled. (hmm, this isn't the right thing to do for rescheduled orders)
- indicate when no items were picked up

SheepTester committed:
exclude fulfilled items from item breakdown

SheepTester committed:
default checkboxes to unchecked

SheepTester committed:
fix fulfilling selected. always indicate whether an item was fulfilled

SheepTester committed:
Show fulfilled badges

SheepTester commented:
I changed the orders view to use badges instead. not sure if it's better

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/e343f844-16a3-4fc3-93f4-09416f84888b)


Marcelo cancelled this rescheduled partially fulfilled order, which is why it has one fulfilled item:

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/ecfca9db-e18d-4fc3-94e3-9d8bba4a9ec0)

alexzhang1618 commented on src/components/admin/store/PickupOrder/index.tsx:
```diff
@@ -0,0 +1,96 @@
+            ) {
+              if (!item.fulfilled) {
+                badge = <span className={styles.notFulfilled}>Not fulfilled</span>;
+              }
```
```suggestion
            if (
              !item.fulfilled && (
                order.status === OrderStatus.FULFILLED || 
                order.status === OrderStatus.PARTIALLY_FULFILLED
              )
            ) {
                badge = <span className={styles.notFulfilled}>Not fulfilled</span>;
```

alexzhang1618 commented on src/components/store/OrderSummary/index.tsx:
```diff
@@ -57,6 +49,12 @@ const OrderItemPreview = ({ item, orderStatus }: OrderItemPreviewProps) => {
+          <div className={styles.badge}>{showNotFulfilled ? 'Picked up' : 'Already picked up'}</div>
+        ) : null}
+        {showNotFulfilled && !item.fulfilled ? (
+          <div className={`${styles.badge} ${styles.notAvailable}`}>Not available</div>
```
Not available makes me think that it's not in stock, can we do something like not picked up maybe? Could also ask others on what wording is best

SheepTester replied:
yeah it should only show for partially fulfilled events, so it wasn't in stock

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/e001e245-16f7-420b-ace4-b9add606165e)

alexzhang1618 commented on src/components/store/OrderSummary/index.tsx:
```diff
@@ -57,6 +49,12 @@ const OrderItemPreview = ({ item, orderStatus }: OrderItemPreviewProps) => {
           fill
         />
+        {showFulfilled && item.fulfilled ? (
+          <div className={styles.badge}>{showNotFulfilled ? 'Picked up' : 'Already picked up'}</div>
```
Thoughts on like a green for this instead of grey similar to how notAvailable is red?

alexzhang1618 reviewed (commented):
I like the tags and the extra text, it looks nice!
Some UI comments:

I think I prefer ✅ and ❌ emojis to the text since it's more concise and readable

I'm also not sure how much I like having all these check boxes to the right of the item. In our store, we don't really have a lot of items where people are ordering a big quantity of them, but maybe we should make this just one check to fulfill all or maybe a number to fulfill a specific number of them? Up to you and whatever looks best
<img width="1440" alt="Screenshot 2024-05-18 at 12 26 44 PM" src="https://github.com/acmucsd/membership-portal-ui-v2/assets/27360825/0f7ca2cb-77fb-403c-9d8a-5c8ccdcd8885">

SheepTester committed:
Labels back to emoji, have checkbox select all of an entire variant at a time

If we want granularity in the future then that can be added later

SheepTester committed:
Merge branch 'main' into sean/fulfill

SheepTester committed:
Display "fulfilled" on fulfilled items in fulfilled/partially fulfilled orders

SheepTester committed:
Make "picked up" badge green

SheepTester commented:
- I switched the text labels in the admin orders page back to emoji
- I merged all the checkboxes into one checkbox per variant. This means that if someone orders multiple ACM AI stickers, but there's somehow only one sticker left, you can only either mark them all as fulfilled or not fulfilled. Maybe that's fine for now.
- I changed the "picked up" / "already picked up" (for rescheduled orders) badge to green:
  
  ![image](https://github.com/user-attachments/assets/ee00d3e9-3dbc-46bd-bfc1-2c5b8d4d45f3)

SheepTester committed:
Merge branch 'main' into sean/fulfill

alexzhang1618 commented on src/components/admin/store/PickupOrder/index.tsx:
```diff
@@ -0,0 +1,102 @@
+              } catch (error: unknown) {
+                reportError('Failed to fulfill order', error);
+              }
+            }}
```
tiny nit but if there's a good amt of code here i'd rather it be a helper function `handleFulfillOrder` so we can keep the jsx nice and tidy

alexzhang1618 reviewed (approved):
LGTM, some nits

Maybe a word break and a little margin between the checkbox n the button?
<img width="1440" alt="Screenshot 2024-08-19 at 11 01 14 PM" src="https://github.com/user-attachments/assets/45483189-4e0d-4938-9cd6-71db6a27fec8">

SheepTester committed:
Address feedback

SheepTester committed:
Merge branch 'sean/fulfill' of https://github.com/acmucsd/membership-portal-ui-v2 into sean/fulfill

SheepTester commented:
![image](https://github.com/user-attachments/assets/2626e925-30a5-4c3b-b1e3-4274d01acde3)
